akka-actor-typed/src/main/scala/akka/actor/typed/internal/ActorContextImpl.scala

```diff
@@ -97,6 +98,19 @@ import akka.util.JavaDurationConverters._
     }(responseTimeout, ClassTag[Res](resClass))
   }
 
+  // Scala API impl
+  def pipeToSelf[Value](future: Future[Value])(mapResult: Try[Value] ⇒ T): Unit = {
+    future.onComplete(value ⇒ self.upcast ! AdaptMessage(value, mapResult))
+  }
+
+  // Java API impl
+  def pipeToSelf[Value](future: CompletionStage[Value], applyToResult: BiFunction[Value, Throwable, T]): Unit = {
+    future whenComplete ((value: Value, ex: Throwable) ⇒ {
+      if (value != null) self ! applyToResult.apply(value, null)
+      if (ex != null) self ! applyToResult.apply(null.asInstanceOf[Value], ex)
```

Comment:
you are fixing the `AdaptMessage` part for the javadsl in another PR?

Reply:
Not myself, but I'm tracking that effort in #25972, if that works for you.